### ISSUE TITLE
Subscriptions are not propagated corrected by indirect leaf node servers, when peer restarts.

### ISSUE THREAD

**Issue by KimonHoffmann** [nats-io/nats-server#2448]
## Defect

When multiple servers form a chain by means of `leafnodes` connections, client subscriptions don't get propagated if a server removed by at least one level of indirection restarts.

Possibly related to #1066

#### Versions of `nats-server` and affected client libraries used:

Verified against:
* `nats-server` version 2.2.6 (macOS) available via Homebrew.
* `nats-server` version 2.1.9 (built from source; reported as 2.2.1-RC12 during startup)

Multiple client libraries consistently show the issue and I was even able to reproduce it locally using `nc`.

#### OS/Container environment: 

Verified on:
* macOS 11.5.2 (amd64)
* Linux 5.4.75 (amd64)
* Linux 5.0.19 (amd64)
 
#### Steps or code to reproduce the issue:

Starting Scenario:
`Client_A` <--> `Server_A` <--> `Server_B` <--> `Server_C` <--> `Client_B`
Where all inter-server connections are **leaf node** connections.

* `Client_A` subscribes to topic `a`.
* `Client_B` publishes to topic `a`, which is correctly received by `Client_A` ✔️ 

Restart `Server_C`

* `Client_B` publishes to topic `a`, which is never received by `Client_A` ❗  

Restart `Server_B`

* `Client_B` publishes to topic `a`, which is again correctly received by `Client_A` ✔️ 

#### Expected result:

After restart of `Server_C` `Client_A` should continue receiving messages published to topic `a`, by `Client_B`.

#### Actual result:

After restart of `Server_C` `Client_A` no longer receives messages published to topic `a`. Restarting the intermediate server `Server_B` fixes the issue.

**Comment by derekcollison** [comment]
Could you test against head of main which is 2.3.5-beta?

**Comment by KimonHoffmann** [comment]
I could reproduce it with the officially released 2.3.4. Will test against HEAD of main later.

**Comment by KimonHoffmann** [comment]
main @ 5d8746e still exhibits the issue.

**Comment by derekcollison** [comment]
ok thanks for that. Will have @matthiashanel take a look.

**Comment by KimonHoffmann** [comment]
I just realized that the title of this issue as well as the steps to reproduce are much more contrived than they would need to be.

The issue is that leaf node servers don't propagate interests that learned about indirectly from another leaf node upstream.

So the reproduction above works just the same without *restarting* `Server_C`, but by just *starting* `Server_C` after `Client_A` has expressed its interest to `Server_A`.

I've updated the title of this issue accordingly.

**Comment by matthiashanel** [comment]
I can't reproduce it. In what I do in the attached gif, do you see anything different to what you do?

![issue-2448](https://user-images.githubusercontent.com/19881388/130502943-a82c3e7c-b6f6-45c3-91c6-2bca0b92e738.gif)

**Comment by KimonHoffmann** [comment]
Interesting, I don't see any *substential* differences to what I'm doing. The only differences I can see right now:
* Ports. I use client listen ports 4222, 4223 & 4224 with leafnode listen ports 7422, 7423 & 7424.
* URL schemes. I use `nats-leaf://`
* Server names. I didn't use explicitly named servers in my reproducer tests above.

I'll investigate this further and will check back.

**Comment by KimonHoffmann** [comment]
Ok, I got it. It's subtle and you almost got me doubting myself. 😅 
It depends to which end of the chain the subscribing entity is connected.

`Client_Pub -> A <-- B <-- C <-- Client_Sub`

Restarting `A` exhibits the problem.

`Client_Sub -> A <-- B <-- C <-- Client_Pub`

Restarting 'C' **does not** exhibit the problem.

**Comment by matthiashanel** [comment]
The gif does show how A is restarted. does it not?

**Comment by KimonHoffmann** [comment]
Yes, but - as noted above - you need to switch the client roles for the issue to show.

The requirement appears to be that:
1. The server (re-)starting is at least one hop removed from the subscribing client and
2. The server (re-)starting is "upstream" from the PoV of the subscribing client, i.e. being connected to instead of initiating the connection.

**Comment by matthiashanel** [comment]
@KimonHoffmann I can verify it now. This is extremely subtle.

**Comment by KimonHoffmann** [comment]
Thanks everyone for the timely triage, fix & bugfix release. Much appreciated!
